log/zapcore_handler.go

```diff
@@ -0,0 +1,129 @@
+package log
+
+import (
+	"fmt"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/log"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"istio.io/pkg/errdict"
+)
+
+var (
+	toLevel = map[zapcore.Level]Level{
+		zapcore.FatalLevel: FatalLevel,
+		zapcore.ErrorLevel: ErrorLevel,
+		zapcore.WarnLevel:  WarnLevel,
+		zapcore.InfoLevel:  InfoLevel,
+		zapcore.DebugLevel: DebugLevel,
+	}
+	toZapLevel = map[Level]zapcore.Level{
+		FatalLevel: zapcore.FatalLevel,
+		ErrorLevel: zapcore.ErrorLevel,
+		WarnLevel:  zapcore.WarnLevel,
+		InfoLevel:  zapcore.InfoLevel,
+		DebugLevel: zapcore.DebugLevel,
+	}
+)
+
+func init() {
+	RegisterDefaultHandler(ZapLogHandlerCallbackFunc)
+}
+
+// ZapLogHandlerCallbackFunc is the handler function that emulates the previous Istio logging output and adds
+// support for errdict package and structured logging.
+func ZapLogHandlerCallbackFunc(
+	level Level,
+	scope *Scope,
+	ie *errdict.IstioErrorStruct,
+	msg string,
+	fields []zapcore.Field) {
+	var ss []string
+	if ie != nil {
+		ss = append(ss, fmt.Sprintf(`"moreInfo":"%s"`, ie.MoreInfo))
+		ss = append(ss, fmt.Sprintf(`"impact":"%s"`, ie.Impact))
+		ss = append(ss, fmt.Sprintf(`"action":"%s"`, ie.Action))
+		ss = append(ss, fmt.Sprintf(`"likelyCauses":"%s"`, ie.LikelyCause))
+	}
+	if len(scope.structuredKey) > 0 {
+		scope.mu.RLock()
```

Comment:
ack, per offline discussion this was to avoid copies but if copies are ok we can get rid of this. i will update with the locks removed and a few more copies.